# Data Cleaning with Pandas

In [1]:
import pandas as pd

## Scenario

As data scientists, we want to build a model to predict the sale price of a house in Seattle in 2019, based on its square footage. We know that the King County Department of Assessments has comprehensive data available on real property sales in the Seattle area. We need to prepare the data.

### First, get the data!

When working on a project involving data that can fit on our computer, we store it in a `data` directory.

```bash
cd <project_directory>  # example: cd ~/flatiron_ds/pandas-3
mkdir data
cd data
```

Note that `<project_directory>` in angle brackets is a _placeholder_. You should type the path to the actual location on your computer where you're working on this project. Do not literally type `<project_directory>` and _do not type the angle brackets_. You can see an example in the _comment_ to the right of the command above.

Now, we'll need to download the two data files that we need. We can do this at the command line:

```bash
wget https://aqua.kingcounty.gov/extranet/assessor/Real%20Property%20Sales.zip
wget https://aqua.kingcounty.gov/extranet/assessor/Residential%20Building.zip
```

*Note:* If you do not have the `wget` command yet, you can install it with `brew install wget`, or use `curl <url> -o <filename>`.

Note that `%20` in a URL translates into a space. Even though you will *never put spaces in filenames*, you may need to deal with spaces that _other_ people have used in filenames.

There are two ways to handle the spaces in these filenames when referencing them at the command line.

#### 1. You can _escape_ the spaces by putting a backslash (`\`, remember _backslash is next to backspace_) before each one:

`unzip Real\ Property\ Sales.zip`

This is what happens if you tab-complete the filename in the terminal. Tab completion is your friend!

#### 2. You can put the entire filename in quotes:

`unzip "Real Property Sales.zip"`

Try unzipping these files with the `unzip` command. The `unzip` command takes one argument, the name of the file that you want to unzip.

In [5]:
sales_df = pd.read_csv('Data/EXTR_RPSale.csv')

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3049: DtypeWarning: Columns (1,2) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Seeing pink? Warnings are useful!

Note the warning above: `DtypeWarning: Columns (1, 2) have mixed types.` Because we start with an index of zero, the columns that we're being warned about are actually the _second_ and _third_ columns, `sales_df['Major']` and `sales_df['Minor']`.

In [6]:
sales_df.head().T

0  \
ExciseTaxNbr                                                  2687551   
Major                                                          138860   
Minor                                                             110   
DocumentDate                                               08/21/2014   
SalePrice                                                      245000   
RecordingNbr                                           20140828001436   
Volume                                                                  
Page                                                                    
PlatNbr                                                                 
PlatType                                                                
PlatLot                                                                 
PlatBlock                                                               
SellerName          WENKLE NOEL SMITH -TRUSTEE                    ...   
BuyerName           ALEXANDER APRIL                               ...   
PropertyType                                                        3   
PrincipalUse                                                        6   
SaleInstrument                                                      3   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                          1   
PropertyClass                                                       8   
SaleWarning                                                             

                                                                    1  \
ExciseTaxNbr                                                  1235111   
Major                                                          664885   
Minor                                                              40   
DocumentDate                                               07/09/1991   
SalePrice                                                           0   
RecordingNbr                                           199203161090     
Volume                                                            071   
Page                                                              001   
PlatNbr                                                        664885   
PlatType                                                            C   
PlatLot                                                B102             
PlatBlock                                                               
SellerName          HERRMAN JEFFREY A+PAULINE A                   ...   
BuyerName           HERRMAN PAULINE A TTEE                        ...   
PropertyType                                                        3   
PrincipalUse                                                        0   
SaleInstrument                                                     26   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                         18   
PropertyClass                                                       3   
SaleWarning                                                        11   

                                                                    2  \
ExciseTaxNbr                                                  2704079   
Major                                                          423943   
Minor                                                              50   
DocumentDate                                               10/11/2014   
SalePrice                                          

### Data overload?

That's a lot of columns. We're only interested in identifying the date, sale price, and square footage of each specific property. What can we do?

In [7]:
sales_df = sales_df[['Major', 'Minor', 'DocumentDate', 'SalePrice']]

In [8]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2026164 entries, 0 to 2026163
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.8+ MB


In [11]:
bldg_df = pd.read_csv('Data/EXTR_ResBldg.csv')

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3049: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Another warning! Which column has index 11?

In [12]:
bldg_df.columns[11]

'ZipCode'

`ZipCode` seems like a potentially useful column. We'll need it to determine which house sales took place in Seattle.

In [13]:
bldg_df.head().T

0                           1  \
Major                                   62304                       62304   
Minor                                    9347                        9378   
BldgNbr                                     1                           1   
NbrLivingUnits                              1                           1   
Address              907  SW 104TH ST   98146  10215   11TH AVE SW  98146   
BuildingNumber                          907                         10215   
Fraction                                                                    
DirectionPrefix                            SW                               
StreetName          104TH                       11TH                        
StreetType                             ST                          AVE      
DirectionSuffix                                                        SW   
ZipCode                                 98146                       98146   
Stories                                     1                           1   
BldgGrade                                   7                           7   
BldgGradeVar                                0                           0   
SqFt1stFloor                             1180                         970   
SqFtHalfFloor                               0                           0   
SqFt2ndFloor                                0                           0   
SqFtUpperFloor                              0                           0   
SqFtUnfinFull                               0                           0   
SqFtUnfinHalf                               0                           0   
SqFtTotLiving                            1570                         970   
SqFtTotBasement                           670                           0   
SqFtFinBasement                           390                           0   
FinBasementGrade                            6                           0   
SqFtGarageBasement                          0                           0   
SqFtGarageAttached                          0                         380   
DaylightBasement                                                        N   
SqFtOpenPorch                             150                           0   
SqFtEnclosedPorch                           0                          80   
SqFtDeck                                    0                           0   
HeatSystem                                  5                           5   
HeatSource                                  2                           2   
BrickStone                                  0                           0   
ViewUtilization                                                         N   
Bedrooms                                    3                           3   
BathHalfCount                               0                           0   
Bath3qtrCount                               0                           1   
BathFullCount                               1                           1   
FpSingleStory                               1                           1   
FpMultiStory                                0                           0   
FpFreestanding                              0                           0   
FpAdditional                                0                           0   
YrBuilt                                  1965                        1969   
YrRenovated                                 0                           0   
PcntComplete                                0                           0   
Obsolescence                                0                           0   
PcntNetCondition                            0                           0   
Condition                                   3                           3   
AddnlCost                                   0                           0   

                                            2                           3  \
Major                                  

### So many features!

As data scientists, we should be _very_ cautious about discarding potentially useful data. But, today, we're interested in _only_ the total square footage of each property. What can we do?


In [14]:
bldg_df = bldg_df[['Major', 'Minor', 'SqFtTotLiving', 'ZipCode']]

In [15]:
bldg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 512114 entries, 0 to 512113
Data columns (total 4 columns):
Major            512114 non-null int64
Minor            512114 non-null int64
SqFtTotLiving    512114 non-null int64
ZipCode          468423 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB


In [16]:
sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])

### Error!

Why are we seeing an error when we try to join the dataframes?

<table>
    <tr>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2013160 entries, 0 to 2013159
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.4+ MB</pre></td>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511359 entries, 0 to 511358
Data columns (total 4 columns):
Major            511359 non-null int64
Minor            511359 non-null int64
SqFtTotLiving    511359 non-null int64
ZipCode          468345 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB
</pre></td>
    </tr>
</table>

Review the error message in light of the above:

* `ValueError: You are trying to merge on object and int64 columns.`

In [17]:
sales_df['Major'] = pd.to_numeric(sales_df['Major'])

ValueError: Unable to parse string "      " at position 949922

### Error!

Note the useful error message above:

`ValueError: Unable to parse string "      " at position 936643`

In this case, we want to treat non-numeric values as missing values. Let's see if there's a way to change how the `pd.to_numeric` function handles errors.

In [4]:
# The single question mark means "show me the docstring"
pd.to_numeric?

Here's the part that we're looking for:
```
errors : {'ignore', 'raise', 'coerce'}, default 'raise'
    - If 'raise', then invalid parsing will raise an exception
    - If 'coerce', then invalid parsing will be set as NaN
    - If 'ignore', then invalid parsing will return the input
```

Let's try setting the `errors` parameter to `'coerce'`.

In [18]:
sales_df['Major'] = pd.to_numeric(sales_df['Major'], errors='coerce')

Did it work?

In [19]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2026164 entries, 0 to 2026163
Data columns (total 4 columns):
Major           float64
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: float64(1), int64(1), object(2)
memory usage: 61.8+ MB


It worked! Let's do the same thing with the `Minor` parcel number.

In [20]:
sales_df['Minor'] = pd.to_numeric(sales_df['Minor'], errors='coerce')

In [21]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2026164 entries, 0 to 2026163
Data columns (total 4 columns):
Major           float64
Minor           float64
DocumentDate    object
SalePrice       int64
dtypes: float64(2), int64(1), object(1)
memory usage: 61.8+ MB


Now, let's try our join again.

In [54]:
sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])

In [44]:
sales_data.size

8677404

In [39]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1446234 entries, 0 to 1446233
Data columns (total 6 columns):
Major            1446234 non-null int64
Minor            1446234 non-null int64
SqFtTotLiving    1446234 non-null int64
ZipCode          1329081 non-null object
DocumentDate     1446234 non-null object
SalePrice        1446234 non-null int64
dtypes: int64(4), object(2)
memory usage: 77.2+ MB


We can see right away that we're missing zip codes for many of the sales transactions. (1321536 non-null entries for ZipCode is fewer than the 1436772 entries in the dataframe.) 

In [25]:
sales_data.loc[sales_data['ZipCode'].isna()].head()

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
91  334330.0  1343.0   05/30/2006          0           4600     NaN
92  334330.0  1343.0   05/30/2006          0           4600     NaN
93  334330.0  1343.0   11/26/2001          0           4600     NaN
94  334330.0  1343.0   05/30/2006          0           4600     NaN
95  334330.0  1343.0   06/30/2016          0           4600     NaN

Because we are interested in finding houses in Seattle zip codes, we will need to drop the rows with missing zip codes.

In [26]:
sales_data = sales_data.loc[~sales_data['ZipCode'].isna(), :]

sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

# Your turn: Data Cleaning with Pandas

### 1. Investigate and drop rows with invalid values in the SalePrice and SqFtTotLiving columns.

Use multiple notebook cells to accomplish this! Press `[esc]` then `B` to create a new cell below the current cell. Press `[return]` to start typing in the new cell.

In [67]:
sales_data.sort_values('SqFtTotLiving').head()

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
515980   941240.0   165.0   12/01/1986     275000              0   98118
646156   917960.0   425.0   06/12/1991      59000              0   98030
1339838  335640.0  1105.0   03/23/2007          0              0   98001
930327   351400.0    25.0   07/23/2018          0              0   98146
330283   356380.0   330.0   03/22/1995     155000              0   98070

In [61]:
sales_data.loc[(sales_data.SalePrice  < 0)]

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
157297   164350.0   77.0   08/29/2017       -600           2280   98133
568561   926491.0  620.0   11/07/2018       -400           2100   98023
1074006  400840.0  155.0   11/21/2017       -300           1830   98058

In [79]:
clean_sales_data=sales_data.loc[(sales_data.SalePrice  >= 0)]

In [77]:
pd.concat([
sales_data.loc[(sales_data.Major == 164350.0) & (sales_data.Minor == 77.0)],
sales_data.loc[(sales_data.Major == 926491.0) & (sales_data.Minor == 620.0)],
sales_data.loc[(sales_data.Major == 400840.0) & (sales_data.Minor == 155.0)]
])

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
157295   164350.0   77.0   12/29/2004     384950           2280   98133
157296   164350.0   77.0   09/23/2009          0           2280   98133
157297   164350.0   77.0   08/29/2017       -600           2280   98133
157298   164350.0   77.0   11/18/2017     655000           2280   98133
568561   926491.0  620.0   11/07/2018       -400           2100   98023
568562   926491.0  620.0   11/07/2018          0           2100   98023
568563   926491.0  620.0   04/30/2019     460000           2100   98023
568564   926491.0  620.0   11/08/2018          0           2100   98023
568565   926491.0  620.0   04/03/2018     336000           2100   98023
568566   926491.0  620.0   03/16/1988          0           2100   98023
1074005  400840.0  155.0   01/24/2013          0           1830   98058
1074006  400840.0  155.0   11/21/2017       -300           1830   98058
1074007  400840.0  155.0   11/13/2000     240000           1830   98058

In [73]:
sales_data.loc[(sales_data.Major == 926491.0) & (sales_data.Minor == 620.0)]

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
568561  926491.0  620.0   11/07/2018       -400           2100   98023
568562  926491.0  620.0   11/07/2018          0           2100   98023
568563  926491.0  620.0   04/30/2019     460000           2100   98023
568564  926491.0  620.0   11/08/2018          0           2100   98023
568565  926491.0  620.0   04/03/2018     336000           2100   98023
568566  926491.0  620.0   03/16/1988          0           2100   98023

In [58]:
sales_data.SalePrice.describe()

count    1.446234e+06
mean     3.174148e+05
std      8.974568e+05
min     -6.000000e+02
25%      0.000000e+00
50%      1.679500e+05
75%      3.699000e+05
max      1.753741e+08
Name: SalePrice, dtype: float64

### 2. Investigate and handle non-numeric ZipCode values

Can you find a way to shorten ZIP+4 codes to the first five digits?

What's the right thing to do with missing values?

In [87]:
sales_data.loc[(len(sales_data.ZipCode) > 5)]

KeyError: True

In [88]:
# Read the error message and decide how to fix it.
# Note: using errors='coerce' is the *wrong* choice in this case.
def is_integer(x):
    try:
        _ = int(x)
    except ValueError:
        return False
    return True

len(sales_data.loc[sales_data['ZipCode'].apply(is_integer) == False, 'ZipCode'])

117421

### 3. Add a column for PricePerSqFt



In [92]:
clean_sales_data = sales_data.loc[(sales_data.SalePrice  >= 0)].copy()

In [94]:
clean_sales_data['PricePerSqFt'] = 0
clean_sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
0  138860.0  110.0   08/21/2014     245000           1490   98002   
1  138860.0  110.0   06/12/1989     109300           1490   98002   
2  138860.0  110.0   01/16/2005      14684           1490   98002   
3  138860.0  110.0   06/08/2005          0           1490   98002   
4  423943.0   50.0   10/11/2014          0            960   98092   

   ProcePerSqFt  
0             0  
1             0  
2             0  
3             0  
4             0

### 4. Subset the data to 2019 sales only.

We can assume that the DocumentDate is approximately the sale date.

In [99]:
clean_sales_data.DocumentDate = pd.to_datetime(clean_sales_data.DocumentDate)
clean_sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1446231 entries, 0 to 1446233
Data columns (total 7 columns):
Major            1446231 non-null float64
Minor            1446231 non-null float64
DocumentDate     1446231 non-null datetime64[ns]
SalePrice        1446231 non-null int64
SqFtTotLiving    1446231 non-null int64
ZipCode          1329078 non-null object
ProcePerSqFt     1446231 non-null int64
dtypes: datetime64[ns](1), float64(2), int64(3), object(1)
memory usage: 88.3+ MB


In [107]:
clean_sales_data.DocumentDate.map(lambda x: x.tm_year()).head()

AttributeError: 'Timestamp' object has no attribute 'tm_year'

### 5. Subset the data to zip codes within the City of Seattle.

You'll need to find a list of Seattle zip codes!

### 6. What is the mean price per square foot for a house sold in Seattle in 2019?

Don't just type the answer. Type code that generates the answer as output!